### PR TITLE
feat(ngMessages): add support for default message

### LINKED ISSUE #12008: ngMessageDefault - Add support for showing values not mapped with ng-message
Many times it is necessary that we show a generic error message for unexpected values.
Would be nice to have something like:

``` HTML
<div ng-messages="controller.errors">
    <div ng-message="error1">Option 1</div>
    <div ng-message="error2">Option 2</div>
    <div ng-message-default>Option 3</div>
</div>
```

This way, whenever there is an "error3" in "controller.errors", as it is not mapped, the value under the ng-message-default would be shown.

### PR DESCRIPTION
**Pull request by gdangelo**
Added support for showing default message when no values are mapped with ng-message.

Closes #12008

[Review on Reviewable](https://reviewable.io/reviews/angular/angular.js/12213)

### CODE REVIEW

**Comment by lgalfaso** [comment]
it looks like this PR adds quite a few of duplicated code. Before doing a proper review, would it be possible to try to keep the core a little bit more DRY?

**Comment by gdangelo** [comment]
I will make the modifications soon.

**Comment by petebacondarwin** [comment]
This looks reasonable let's get it into 1.6 early as soon as 1.5.0 is released.

**Comment by mikeycgto** [comment]
Any updates on this making it into 1.6?

File: src/ngMessages/messages.js [hunk 1]
```diff
@@ -239,6 +239,19 @@ var jqLite = angular.element;
...
+ * ## Displaying a default message
+ * If the ngMessages renders no inner ngMessage directive (that is to say when the key values does not
+ * match the attribute value present on each ngMessage directive), then it will render a default message
+ * using the ngMessageDefault directive.
+ *
+ * ```html
+ * <div ng-messages="myForm.myField.$error" role="alert">
+ *   <div ng-message="required">This field is required</div>
+ *   <div ng-message="minlength">This field is too short</div>
+ *   <div ng-message-default>This is a default message</div>
```
Comment (gkalpak) [inline src/ngMessages/messages.js]: I would use a more meaningful message here to give a better impression of how to use the directive.

File: src/ngMessages/messages.js [hunk 2]
```diff
@@ -379,9 +398,17 @@ angular.module('ngMessages', [])
              messageCtrl.detach();
            });
 
-           unmatchedMessages.length !== totalMessages
-              ? $animate.setClass($element, ACTIVE_CLASS, INACTIVE_CLASS)
-              : $animate.setClass($element, INACTIVE_CLASS, ACTIVE_CLASS);
+           if (unmatchedMessages.length !== totalMessages) {
+             // Unset default message if setted
```
Comment (gkalpak) [inline src/ngMessages/messages.js]: _setted_  -->  _set_

File: src/ngMessages/messages.js [hunk 3]
```diff
@@ -397,23 +424,32 @@ angular.module('ngMessages', [])
...
-           var nextKey = latestKey.toString();
-           messages[nextKey] = {
-             message: messageCtrl
-           };
-           insertMessageNode($element[0], comment, nextKey);
-           comment.$$ngMessageNode = nextKey;
-           latestKey++;
+         this.register = function(comment, messageCtrl, isDefault) {
+           if (isDefault) {
+            ctrl.default = messageCtrl;
```
Comment (gkalpak) [inline src/ngMessages/messages.js]: Wrong indentation (here and below).

File: src/ngMessages/messages.js [hunk 4]
```diff
@@ -397,23 +424,32 @@ angular.module('ngMessages', [])
...
          };
 
-         this.deregister = function(comment) {
-           var key = comment.$$ngMessageNode;
-           delete comment.$$ngMessageNode;
-           removeMessageNode($element[0], comment, key);
-           delete messages[key];
+         this.deregister = function(comment, isDefault) {
+           if (isDefault) {
+            delete ctrl.default;
```
Comment (gkalpak) [inline src/ngMessages/messages.js]: I think you also need to detach `ctrl.default` (if defined), because if it is visible, it won't get removed by `ctrl.reRender()`.

File: src/ngMessages/messages.js [hunk 5]
```diff
@@ -594,35 +630,74 @@ angular.module('ngMessages', [])
...
+    *
+    * <!-- or by using element directives -->
+    * <ng-messages for="expression" role="alert">
+    *   <ng-message when="stringValue">...</ng-message>
+    *   <ng-message when="stringValue1, stringValue2, ...">...</ng-message>
+    *   <ng-message-default>...</ng-message-default>
+    * </ng-messages>
+    * ```
+    *
+    * @param {expression} ngMessageDefault|when no ngMessage matches.
```
Comment (gkalpak) [inline src/ngMessages/messages.js]: It doesn't take any params.

File: test/ngMessages/messagesSpec.js [hunk 6]
```diff
@@ -401,6 +401,24 @@ describe('ngMessages', function() {
...
   });
 
+  it('should render a default message when no one matched', inject(function($rootScope, $compile) {
+    element = $compile('<div ng-messages="col">' +
+                       '  <div ng-message="val">Message is set</div>' +
+                       '  <div ng-message-default>Default message is set</div>' +
+                       '</div>')($rootScope);
+    $rootScope.$digest();
+
+    //expect(element.text()).not.toContain('Message is set');
```
Comment (gkalpak) [inline test/ngMessages/messagesSpec.js]: Remove comment.

File: test/ngMessages/messagesSpec.js [hunk 7]
```diff
@@ -401,6 +401,24 @@ describe('ngMessages', function() {
...
+    expect(element.text()).toContain('Default message is set');
```
Comment (gkalpak) [inline test/ngMessages/messagesSpec.js]: Why not use `.toBe()`?

File: test/ngMessages/messagesSpec.js [hunk 8]
```diff
@@ -401,6 +401,24 @@ describe('ngMessages', function() {
...
+
+    $rootScope.$apply(function() {
+      $rootScope.col = { val: true };
+    });
```
Comment (gkalpak) [inline test/ngMessages/messagesSpec.js]: `$rootScope.$apply('col = {val: true}')` is more concise :smiley:

File: test/ngMessages/messagesSpec.js [hunk 9]
```diff
@@ -401,6 +401,24 @@ describe('ngMessages', function() {
...
+
+    //expect(element.text()).toContain('Message is set');
+    expect(element.text()).not.toContain('Default message is set');
```
Comment (gkalpak) [inline test/ngMessages/messagesSpec.js]: Again, why not use `.toBe()` and verify the exact contents?

**Comment by gkalpak** [comment]
A few more tests would be (e.g. using `ng-messages-multiple` or using `ngIf` on `ngMessageDefault` etc). But overall looks good.

File: src/ngMessages/messages.js (hunk 4 above)
Comment (mikeycgto) [inline src/ngMessages/messages.js]: Do you mean to call `removeMessageNode` here for the default?

File: src/ngMessages/messages.js (hunk 2 above)
Comment (mikeycgto) [inline src/ngMessages/messages.js]: Not sure I follow...?

File: src/ngMessages/messages.js (hunk 1 above)
Comment (mikeycgto) [inline src/ngMessages/messages.js]: Anything in specific that should be mentioned here?

Comment (gkalpak) [inline src/ngMessages/messages.js]: Anything that you would use in an actual app. In this case, since these seem to be validation messages, the default message would be something like 'Valid' or a checkmark or 'This field is valid` etc.

File: src/ngMessages/messages.js (hunk 2 above)
Comment (gkalpak) [inline src/ngMessages/messages.js]: `X  -->  Y` is my way of telling that there is a typo in `X`, it should be changed to `Y` :stuck_out_tongue: 
(Basically, change `setted` to `set`.)

File: src/ngMessages/messages.js (hunk 4 above)
Comment (gkalpak) [inline src/ngMessages/messages.js]: I don't think `removeMessageNode()` works for `ctrl.default`. I was thinking `ctrl.default.detach()`.
<sub>And there should also be a test about dynamically adding/removing an `<ng-message-default>` element (e.g. using `ngIf`).</sub>

**Comment by gdangelo** [comment]
Hi @gkalpak, @mikeycgto, thanks to follow this feature request. Do we need anything else to include the feature in the 1.6 release?

**Comment by petebacondarwin** [comment]
@gdangelo - yes, @gkalpak provided a number of comments that need addressing, including some more tests.
